Reworked cleanroom structure check.

AbdielKavash opened the pull request:
This PR completely reworks the structure check and requirements for the cleanroom, in order to make the rules much simpler to understand for the player, easier to check even by hand, and less prone to coding errors.

Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17523.
Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17341.
Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/16809.
Fixes some miscellaneous problems which don't have posted issues.

Full details of the new structure requirements:

* Cleanroom footprint must still be a rectangle. It is no longer limited to odd dimensions; in case the width or length is even the controller needs to be one of the center two (or four) blocks.
* Edges of the top layer must be plascrete, all other blocks must be filter casings (this is unchanged).
* Minimum height is 4 (unchanged).
* Maximum width and height are now both configurable; default to 15 (same as before).
* There is a hard minimum number of plascrete blocks, which is now configurable, and defaults to 20 (same as before).
* Check for non-plascrete blocks was completely redone. The only rule now is that a certain percentage of plascrete in walls and floor can be replaced by any other valid block. This percentage is also configurable, and defaults to 30%. All replacement blocks now count towards this single limit, there are no per-block fixed or relative limits any more.
* Note that this percentage now includes functional blocks, like hatches or diodes. Therefore it should not be pushed *too* low. Still it allows for at least one side of the CR to be made mostly transparent for vision.
* Any plascrete except for the edges of the top layer can be replaced by other valid blocks.
* Reinforced Doors now have to be properly "sealed". Adjacent doors on different sides of the block will now result in the CR losing efficiency, as if the door was open. Doors also can not replace the vertical edges of the CR. Finally, since all replacement blocks now count towards the same limit, there is no longer a maximum cap on 2 doors -- one can put as many as the limit allows.
* The list of valid replacement blocks is still configurable, although the format of the config file was changed (simplified, due to the removal of per-block limits).
* Any EV+ tier glass can now be used as a replacement block. (This is in addition to other glass types, such as Warded Glass, which are still allowed.)
* Ethereal Glass from Extra Utilities has been added as a valid replacement block. This requires unstable ingots to craft, and is intended as a replacement for the door configuration which used to allow the player to walk inside the CR without losing efficiency.
* Added a warning in the tooltip about machines voiding outputs when the CR is below 100% efficiency. I know this has been getting people every now and then.

Several issues have been fixed:
* An issue where the northeast vertical edge was used to determine height of the cleanroom. This meant that that specific edge had to be pure plascrete, and no other blocks; and also that an extra plascrete block below the cleanroom could confuse the height measurement and result in an invalid cleanroom. Both have been fixed, as the height is now calculated differently.
* An issue where the player could place two reinforced doors in a configuration with a gap between them to move in and out of the cleanroom without having to open them. Status of doors is now checked more carefully, and such configurations will result in the CR losing efficiency as if the door was opened. (Though it still remains a valid structure.)
* Recently introduced issues where certain blocks were either incorrectly not recognized, or not recognized when placed in some positions or some combinations.

Potential issues:
* This change does invalidate some cleanroom structures which were valid as of 2.7.0-beta-2, particularly those which push the limit of combining different glass types. I am open to tuning the 30% replacement limit up or down further (full disclosure, my preferred number would be up somewhere around 50%, but other devs wanted it lower.)
* If anybody feels *really* strongly that multiple elevators/travel anchors/etc. should absolutely not be allowed, please explain your reasoning. I see no practical reason to limit this and add unnecessary complexity to the checks.

My own questions:
* Since I have changed the config file format, should I also delete the old configs, if found? Currently the config file is being generated correctly in a new installation; however, if one updates GT5 and keeps their old config file, both old and new options will remain (the old ones are ignored). This could be potentially confusing.
* It was suggested on discord by @nicksitnikov that the edge of the top layer of the CR should also be able to be replaced (mostly to house hatches, etc.) This could be changed without too much effort; although I personally don't find it as aesthetical as forcing the plascrete edge. Does anyone have an opinion on this?
* @mike_be on discord also suggested being able to rotate the cleanroom controller, which would put the filter casing layer on some other side than the top. It would certainly be possible... thoughts?

TODO: See if any quests need to be updated to account for the changes.

<details>
<summary>Tooltip screenshots</summary>

![](https://i.imgur.com/ROKvQ3R.png)
![](https://i.imgur.com/pQ4mrrm.png)

Even-width valid cleanroom:
![](https://i.imgur.com/wU3nDUU.png)
</details>

boubou19 commented on src/main/java/gregtech/common/tileentities/machines/multi/MTECleanroom.java:
```diff
@@ -449,79 +744,78 @@ public void construct(ItemStack itemStack, boolean b) {
-            this.allowedCount = count;
-        }
-    }
+    public static void loadConfig(Configuration cfg) {
```
this should be removed and migrated to gtnhlib config format. Either put it in a new file or put it there: https://github.com/GTNewHorizons/GT5-Unofficial/blob/master/src/main/java/gregtech/common/config/MachineStats.java

AbdielKavash replied:
Moved to `MachineStats`. Removed config from max size.

boubou19 commented:
i'm not a fan of configurable maximum size of the multi. the default values are more than enough for maximum size. After either use multis or make a second one. I have never felt the need to have more than one max cleanroom with a quantum ring and 40 zpm circuit assemblers in my playthrough.

Breaking setup with such structure change is unavoidable. So if you think only a tiny portion of the CRs will break, that's acceptable if we compare to what freedom we give to the player.

Multiple elevators/traveler anchor is weird, but fine. Nothing game breaking there, when you can cheese 30% of the plascrete blocks with glass anyways.

about configs, they must be coded with pack defaults and a PR to the modpack repo is needed to update the configs. Forge will anyways disregard configs that are left over in the config files.

> It was suggested on discord by @nicksitnikov that the edge of the top layer of the CR should also be able to be replaced (mostly to house hatches, etc.) This could be changed without too much effort; although I personally don't find it as aesthetical as forcing the plascrete edge. Does anyone have an opinion on this?

Aside maintenance hatch and energy hatch, rest does not make sense, so you could lift the restriction for those i guess.

> @mike_be on discord also suggested being able to rotate the cleanroom controller, which would put the filter casing layer on some other side than the top. It would certainly be possible... thoughts?

why not, but not in this PR, this is an important multi, let's not introduce more testing cases. We must ensure your code changes works for a huge variety of combinations

AbdielKavash commented:
> i'm not a fan of configurable maximum size of the multi. the default values are more than enough for maximum size. After either use multis or make a second one. I have never felt the need to have more than one max cleanroom with a quantum ring and 40 zpm circuit assemblers in my playthrough.

Removed.

> > It was suggested on discord by @nicksitnikov that the edge of the top layer of the CR should also be able to be replaced (mostly to house hatches, etc.) This could be changed without too much effort; although I personally don't find it as aesthetical as forcing the plascrete edge. Does anyone have an opinion on this?
> 
> Aside maintenance hatch and energy hatch, rest does not make sense, so you could lift the restriction for those i guess.

Made it so that the top layer can be replaced as well. It also now counts towards the limits (both 20 required blocks and 30% replaceable plascrete blocks), which is probably more intuitive to the player.

I also simplified the logic and patched some more edge cases (no more doors in the floor). Hopefully the code should be easy to follow for future editors.
